File: sdk/basin-js/test/storage.test.ts
import { describe, expect, it } from "vitest";

import { createClient } from "../src/index.js";
import { FetchMock } from "./helpers.js";

const URL_BASE = "http://basin.test";
const PROJECT = "01JTESTPROJECT0000000000";

function client(fetch: FetchMock) {
  return createClient(URL_BASE, "key", { projectId: PROJECT, fetch: fetch.impl });
}

describe("storage", () => {
  it("createBucket POSTs /storage/v1/bucket with the documented body", async () => {
    const fetch = new FetchMock().queueJson(
      {
        id: "b1", name: "avatars", public: true, file_size_limit: null,
        allowed_mime_types: ["image/png"], created_at: "t", updated_at: "t",
      },
      201,
    );
    const bucket = await client(fetch).storage.createBucket("avatars", {
      public: true,
      allowedMimeTypes: ["image/png"],
    });
    expect(bucket.name).toBe("avatars");
    expect(fetch.last.url).toBe(`${URL_BASE}/storage/v1/bucket`);
    expect(JSON.parse(fetch.last.body!)).toEqual({
      name: "avatars",
      public: true,
      file_size_limit: null,
      allowed_mime_types: ["image/png"],
    });
  });

  it("getBucket / deleteBucket bind /storage/v1/bucket/:name", async () => {
    const fetch = new FetchMock()
      .queueJson({ id: "b1", name: "avatars", public: false, file_size_limit: null, allowed_mime_types: [], created_at: "t", updated_at: "t" })
      .queueJson({ message: "bucket deleted", objects_purged: 3 });
    const c = client(fetch);
    await c.storage.getBucket("avatars");
    expect(fetch.calls[0]!.method).toBe("GET");
    expect(fetch.calls[0]!.url).toBe(`${URL_BASE}/storage/v1/bucket/avatars`);

    const res = await c.storage.deleteBucket("avatars");
    expect(res.objects_purged).toBe(3);
    expect(fetch.calls[1]!.method).toBe("DELETE");
  });

  it("upload POSTs raw bytes with content-type to /storage/v1/object/:bucket/*path", async () => {
    const fetch = new FetchMock().queueJson({
      id: "o1", bucket_id: "b1", path: "a/b.png", size: 3, mime_type: "image/png",
      metadata: null, owner: null, etag: "abc", created_at: "t", updated_at: "t",
    });
    const obj = await client(fetch)
      .storage.from("avatars")
      .upload("a/b.png", new Uint8Array([1, 2, 3]), { contentType: "image/png" });
    expect(obj.etag).toBe("abc");
    expect(fetch.last.url).toBe(`${URL_BASE}/storage/v1/object/avatars/a/b.png`);
    expect(fetch.last.method).toBe("POST");
    expect(fetch.last.headers["content-type"]).toBe("image/png");
  });

  it("download GETs the object and returns bytes + headers", async () => {
    const fetch = new FetchMock().queue(
      new Response(new Uint8Array([9, 9]), {
        status: 200,
        headers: { "content-type": "image/png", etag: "abc" },
      }),
    );
    const res = await client(fetch).storage.from("avatars").download("a/b.png");
    expect(fetch.last.method).toBe("GET");
    expect(new Uint8Array(res.data)).toEqual(new Uint8Array([9, 9]));
    expect(res.contentType).toBe("image/png");
    expect(res.etag).toBe("abc");
  });

  it("list POSTs prefix/limit/offset to /storage/v1/object/list/:bucket", async () => {
    const fetch = new FetchMock().queueJson([]);
    await client(fetch).storage.from("avatars").list({ prefix: "a/", limit: 10 });
    expect(fetch.last.url).toBe(`${URL_BASE}/storage/v1/object/list/avatars`);
    expect(fetch.last.method).toBe("POST");
    expect(JSON.parse(fetch.last.body!)).toEqual({ prefix: "a/", limit: 10, offset: null });
  });

  it("removeByPrefixes sends DELETE with a prefixes[] body", async () => {
    const fetch = new FetchMock().queueJson({ deleted: 2 });
    const res = await client(fetch).storage.from("avatars").removeByPrefixes(["a/"]);
    expect(res.deleted).toBe(2);
    expect(fetch.last.url).toBe(`${URL_BASE}/storage/v1/object/avatars`);
    expect(fetch.last.method).toBe("DELETE");
    expect(JSON.parse(fetch.last.body!)).toEqual({ prefixes: ["a/"] });
  });

  it("getPublicUrl embeds the project id (no-JWT route shape)", () => {
    const url = client(new FetchMock()).storage.from("avatars").getPublicUrl("a/b.png");
    expect(url).toBe(
      `${URL_BASE}/storage/v1/object/public/${PROJECT}/avatars/a/b.png`,
    );
  });

  it("createSignedUrl POSTs sign/upload and returns the camelCase body", async () => {
    const fetch = new FetchMock().queueJson({
      signedUrl: `/storage/v1/object/sign/${PROJECT}/avatars/a/b.png?token=aa&expires=99`,
      expiresAt: "2026-01-01T00:00:00Z",
    });
    const signed = await client(fetch)
      .storage.from("avatars")
      .createSignedUrl("a/b.png", { expiresIn: 600 });
    expect(fetch.last.url).toBe(
      `${URL_BASE}/storage/v1/object/sign/upload/avatars/a/b.png`,
    );
    expect(JSON.parse(fetch.last.body!)).toEqual({ expires_in: 600 });
    expect(signed.expiresAt).toBe("2026-01-01T00:00:00Z");
    expect(signed.absoluteUrl).toBe(
      `${URL_BASE}/storage/v1/object/sign/${PROJECT}/avatars/a/b.png?token=aa&expires=99`,
    );
  });
});

File: sdks/dart/test/storage_test.dart
import 'dart:convert';

import 'package:basin_sdk/basin_sdk.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

BasinClient _client(http.Client mock) => BasinClient.create(
  url: 'http://localhost:8080',
  key: 'k',
  projectId: 'proj-01',
  httpClient: mock,
);

final _bucketJson = {
  'id': 'b1',
  'name': 'avatars',
  'public': false,
  'file_size_limit': null,
  'allowed_mime_types': [],
  'created_at': '2026-01-01T00:00:00Z',
  'updated_at': '2026-01-01T00:00:00Z',
};

final _objectJson = {
  'id': 'o1',
  'bucket_id': 'b1',
  'path': 'users/alice.png',
  'size': 1024,
  'mime_type': 'image/png',
  'metadata': null,
  'owner': null,
  'etag': '"abc"',
  'created_at': '2026-01-01T00:00:00Z',
  'updated_at': '2026-01-01T00:00:00Z',
};

void main() {
  group('StorageClient.createBucket', () {
    test('POSTs to /storage/v1/bucket', () async {
      http.Request? req;
      final mock = MockClient((r) async {
        req = r;
        return http.Response(jsonEncode(_bucketJson), 201,
            headers: {'content-type': 'application/json'});
      });
      final client = _client(mock);
      final bucket = await client.storage.createBucket(
        'avatars',
        public: false,
        allowedMimeTypes: ['image/png'],
      );
      expect(req?.method, 'POST');
      expect(req?.url.path, '/storage/v1/bucket');
      final body = jsonDecode(req!.body) as Map;
      expect(body['name'], 'avatars');
      expect(body['allowed_mime_types'], ['image/png']);
      expect(bucket.name, 'avatars');
    });
  });

  group('StorageClient.getBucket', () {
    test('GETs /storage/v1/bucket/:name', () async {
      http.Request? req;
      final mock = MockClient((r) async {
        req = r;
        return http.Response(jsonEncode(_bucketJson), 200,
            headers: {'content-type': 'application/json'});
      });
      final client = _client(mock);
      await client.storage.getBucket('avatars');
      expect(req?.url.path, '/storage/v1/bucket/avatars');
    });
  });

  group('StorageBucketClient.upload', () {
    test('POSTs raw bytes to /storage/v1/object/:bucket/*path', () async {
      http.Request? req;
      final mock = MockClient((r) async {
        req = r;
        return http.Response(jsonEncode(_objectJson), 200,
            headers: {'content-type': 'application/json'});
      });
      final client = _client(mock);
      final data = [0x89, 0x50, 0x4e, 0x47]; // PNG header
      await client.storage.fromBucket('avatars').upload(
        'users/alice.png',
        data,
        contentType: 'image/png',
      );
      expect(req?.url.path, '/storage/v1/object/avatars/users/alice.png');
      expect(req?.method, 'POST');
      expect(req?.headers['Content-Type'], 'image/png');
      expect(req?.bodyBytes, data);
    });
  });

  group('StorageBucketClient.download', () {
    test('GETs /storage/v1/object/:bucket/*path', () async {
      final mock = MockClient((_) async => http.Response.bytes(
        [1, 2, 3],
        200,
        headers: {'content-type': 'image/png', 'etag': '"abc"'},
      ));
      final client = _client(mock);
      final result =
          await client.storage.fromBucket('avatars').download('users/alice.png');
      expect(result.data, [1, 2, 3]);
      expect(result.contentType, 'image/png');
      expect(result.etag, '"abc"');
    });
  });

  group('StorageBucketClient.list', () {
    test('POSTs to /storage/v1/object/list/:bucket', () async {
      http.Request? req;
      final mock = MockClient((r) async {
        req = r;
        return http.Response(jsonEncode([_objectJson]), 200,
            headers: {'content-type': 'application/json'});
      });
      final client = _client(mock);
      final objs = await client.storage.fromBucket('avatars').list(
        prefix: 'users/',
        limit: 10,
      );
      expect(req?.url.path, '/storage/v1/object/list/avatars');
      expect(req?.method, 'POST');
      final body = jsonDecode(req!.body) as Map;
      expect(body['prefix'], 'users/');
      expect(objs, hasLength(1));
    });
  });

  group('StorageBucketClient.removeByPrefixes', () {
    test('DELETEs /storage/v1/object/:bucket with prefixes body', () async {
      http.Request? req;
      final mock = MockClient((r) async {
        req = r;
        return http.Response(jsonEncode({'ok': true}), 200,
            headers: {'content-type': 'application/json'});
      });
      final client = _client(mock);
      await client.storage.fromBucket('avatars').removeByPrefixes(['users/']);
      expect(req?.url.path, '/storage/v1/object/avatars');
      expect(req?.method, 'DELETE');
      final body = jsonDecode(req!.body) as Map;
      expect(body['prefixes'], ['users/']);
    });
  });

  group('StorageBucketClient.getPublicUrl', () {
    test('builds URL with project_id in path', () {
      final client = BasinClient.create(
        url: 'https://api.basin.run',
        projectId: 'proj-01',
        httpClient: MockClient((_) async => http.Response('{}', 200)),
      );
      final url =
          client.storage.fromBucket('avatars').getPublicUrl('users/alice.png');
      expect(url,
          'https://api.basin.run/storage/v1/object/public/proj-01/avatars/users/alice.png');
    });

    test('throws E_INVALID_REQUEST when project_id unknown', () {
      final client = BasinClient.create(
        url: 'https://api.basin.run',
        httpClient: MockClient((_) async => http.Response('{}', 200)),
      );
      expect(
        () => client.storage.fromBucket('avatars').getPublicUrl('file.png'),
        throwsA(isA<BasinApiError>().having(
          (e) => e.code,
          'code',
          'E_INVALID_REQUEST',
        )),
      );
    });
  });

  group('StorageBucketClient.createSignedUrl', () {
    test('POSTs to sign/upload path with expires_in', () async {
      http.Request? req;
      final mock = MockClient((r) async {
        req = r;
        return http.Response(
          jsonEncode({
            'signedUrl': '/storage/v1/object/sign/proj-01/avatars/file.png?token=tok',
            'expiresAt': '2026-06-01T01:00:00Z',
          }),
          200,
          headers: {'content-type': 'application/json'},
        );
      });
      final client = _client(mock);
      final su = await client.storage
          .fromBucket('avatars')
          .createSignedUrl('file.png', expiresIn: 7200);
      expect(req?.url.path,
          '/storage/v1/object/sign/upload/avatars/file.png');
      final body = jsonDecode(req!.body) as Map;
      expect(body['expires_in'], 7200);
      expect(su.expiresAt, '2026-06-01T01:00:00Z');
      expect(su.absoluteUrl, startsWith('http://localhost:8080'));
    });
  });

  group('path encoding', () {
    test('spaces and special chars in object path are percent-encoded', () async {
      http.Request? req;
      final mock = MockClient((r) async {
        req = r;
        return http.Response(jsonEncode(_objectJson), 200,
            headers: {'content-type': 'application/json'});
      });
      final client = _client(mock);
      await client.storage
          .fromBucket('my bucket')
          .download('my file/with spaces.txt');
      // Bucket name encoded in path
      expect(req?.url.path, contains('my%20bucket'));
      // Path segment encoded
      expect(req?.url.path, contains('my%20file'));
    });
  });
}
